python/ray/tune/experimental/output.py

```diff
@@ -193,26 +193,32 @@ def _get_trials_with_error(trials: List[Trial]) -> List[Trial]:
     return [t for t in trials if t.error_file]
 
 
-def _infer_user_metrics(trials: List[Trial], limit: int = 4) -> List[str]:
+def _infer_user_metrics(
+    trials: List[Trial], limit: int = 5, col_mapping: Optional[Dict[str, str]] = None
+) -> List[str]:
     """Try to infer the metrics to print out.
 
     By default, only the first 4 meaningful metrics in `last_result` will be
     inferred as user implied metrics.
     """
+    col_mapping = col_mapping or {}
     # Using OrderedDict for OrderedSet.
     result = collections.OrderedDict()
     for t in trials:
         if not t.last_result:
             continue
-        for metric, value in t.last_result.items():
-            if metric not in DEFAULT_COLUMNS:
-                if metric not in AUTO_RESULT_KEYS:
-                    if type(value) in VALID_SUMMARY_TYPES:
-                        result[metric] = ""  # not important
+        for flat_metric, value in flatten_dict(t.last_result).items():
```

Comment:
Cleaned this up a little.

Reply:
We have to flatten here, b/c last results (and also `DEFAULT_COLUMNS` or `AUTO_RESULT_KEYS`) might contain nested keys.